src/Disciplina/Educator/Validate.hs

```diff
@@ -0,0 +1,32 @@
+
+-- | Validation of private block
+
+module Disciplina.Educator.Validate
+       ( validate
+       ) where
+
+import Universum
+
+import Disciplina.Educator.Block (PrivateBlock, pbHeader, pbBody, pbbTxs, pbhBodyProof)
+import Disciplina.Educator.Txs (PrivateTxAux (..), PrivateTxWitness (..))
+import Disciplina.Educator.Serialise ()
+import Disciplina.Crypto.MerkleTree (fromFoldable, getMerkleRoot)
+import Disciplina.Crypto (hash, verify)
+
+-- | Validate private block.
+-- Block is valid if header matches body
+-- and all transactions have valid signatures.
+validate :: PrivateBlock -> Bool
+validate pb =
+    let txs = pb ^. pbBody . pbbTxs
+        merkleRoot = getMerkleRoot (fromFoldable txs)
+        headerProof = pb ^. pbHeader . pbhBodyProof
+        headerProofValid = headerProof == merkleRoot
+        txsSigProofValid = all (==True) (map validateTx txs)
+    in headerProofValid && txsSigProofValid
+  where validateTx :: PrivateTxAux -> Bool
+        validateTx (PrivateTxAux {..}) =
+          let txProofSig = _ptwSig _ptaWitness
+              txProofKey = _ptwKey _ptaWitness
+              tx = hash _ptaTx
+          in verify txProofKey tx txProofSig
```

Comment:
You forgot that `PublicKey` from witness should correspond to `StudentId` (i. e. `StudentId` should be an `Address` derived from `PublicKey`.